tests/cbmc/proofs/s2n_connection_get_last_message_name/s2n_connection_get_last_message_name_harness.c

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <tls/s2n_connection.h>
+
+void s2n_connection_get_last_message_name_harness()
+{
+    /* Non-deterministic inputs. */
+    struct s2n_connection *s2n_connection = malloc(sizeof(*s2n_connection));
+
+    /* Operation under verification. */
+    const char* last_message_name = s2n_connection_get_last_message_name(s2n_connection);
+
+    /* Post-conditions. */
+    assert(S2N_IMPLIES(last_message_name != NULL, s2n_result_is_ok(s2n_handshake_validate(&(s2n_connection->handshake)))));
```

Comment:
The post condition looks flipped.

We are checking validity of the handshake if the return value isn't `NULL`.
Why not check that the function returns non-NULL values for valid handshakes? 
Better yet, consider using `S2N_IFF` in place of `S2N_IMPLIES` here to check both implications?

Reply:
> We are checking validity of the handshake if the return value isn't `NULL`.

Yes.
> Why not check that the function returns non-NULL values for valid handshakes?

We're relying on short circuiting. This macro will be expanded to `(!(a) || (b))`, which means that condition `a` will always be evaluated. If we check the validity of `handshake` first, we may fall into the case where `s2n_connection` is NULL, which would lead to NULL pointer dereferences.

> Better yet, consider using `S2N_IFF` in place of `S2N_IMPLIES` here to check both implications?

Same reason as above. We can not access `handshake` in the case `s2n_connection` is NULL.

Reply:
> This macro will be expanded to `(!(a) || (b))`, which means that condition `a` will always be evaluated.

That doesn't matter. `a` being always evaluated doesn't really justify correctness of the postcondition. My main concern is that if `s2n_connection_get_last_message_name` is unsound and _always_ returns `NULL`, then the implication in the postcondition is vacuously `true` always, and it kind of useless.

> If we check the validity of handshake first, we may fall into the case where `s2n_connection` is NULL, which would lead to `NULL` pointer dereferences.

How about the following postcondition:

```suggestion
    assert(S2N_IMPLIES(
      s2n_connection != NULL && s2n_result_is_ok(s2n_handshake_validate(&(s2n_connection->handshake))),
      last_message_name != NULL
    ));
```

Reply:
Good catch! It'd be in fact vacuously true if somebody change the implementation to always return NULL.